### PR TITLE
Add more options for discontinued products

### LINKED ISSUE #23533: Add an option to keep inactive products enabled
## Is your feature request related to a problem?

When you disable a product, these are options of we can do with them.
- 404
- 410
- 301 product redirect
- 302 product redirect
- 301 category redirect
- 302 category redirect

None of these options tell the customer what happened, why they cannot reach the page or why they are not shown the product they clicked on on google. 404 is something like a plague, it should be avoided at all costs and incredibly hurts the site's SEO.

## Describe the solution you'd like

Add an option not to return anything and just display the page. This is done by Amazon, eBay and all the big sellers.
User would land on the product page, see add-to-cart disabled and he would know what happened and would not be confused.
We can also display other products from the default category of this product.

The modification would not be difficult and would require only small tweak in ProductController and the template.

## Design proposition

Výstřižek

### 1. New options on product page
- 200 (keep product displayed as discontinued)
- 404 (keep product displayed as discontinued)
- 410 (keep product displayed as discontinued)
- default (as set in prestashop settings)

### 2. Helper text under these options
Snímek obrazovky 2022-09-22 131014

### 3. New option in prestashop settings called `Default behavior for disabled products`. You could select a default state that all products will have. Options:
- 404
- 410
- 200 (keep product displayed as discontinued)
- 404 (keep product displayed as discontinued)
- 410 (keep product displayed as discontinued)

```sql
ALTER TABLE `ps_product` CHANGE `redirect_type` `redirect_type` ENUM('404','410','301-product','302-product','301-category','302-category','200-displayed','404-displayed','410-displayed','default') CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci NOT NULL DEFAULT 'default';
```

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See description below
| Type?             | new feature
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #23533
| Related PRs       | https://github.com/PrestaShop/autoupgrade/pull/538
| How to test?      | -
| Possible impacts? | -

# Description
### Improved configuration
- Currently, every product had a hardcoded 404 redirection settings by default.
- If you wanted to change behavior for all products, you would have to run DB query or you click yourself to death.
- This PR adds a new field into Prestashop settings, where you can configure this behavior. Also, every new product has this new `default` type.
- You can then customize the settings per product as you are used to.

default

### New options on what to do with discontinued products
- This PR adds a possibility to keep discontinued products disabled with "SALE ENDED" message.
- The products have proper structured data availability `Discontinued`, a nice message that it's no longer in sale and disabled add to cart button.
- You are able to set your desired HTTP response depending on your SEO strategy. Some people want to keep sending 200 and keep products indexed. Some people would like to send 410 and tell google to remove this product from index. 404 is also an option, although I would not recommend using it.

### Technical
- Works on old and new product page perfectly.
- New functions of this PR work standalone even without a template modification.
- Products with empty values in database will work fine and use the default value from configuration. Even if there was some nonsense in the database like custom ENUMs, it will still fall back on 404 at least in the switch.
- It can be made more beautiful by just using `{if $product.active != 1}` on some places in your template, but that's up to every shop owner on what he wants to hide on discontinued products or how to customize the product page in that case.

# How does it look and work
This is a template without any modification:
no template mod
Proper response if wanted:
response
Proper structured data:
structured

# How to test
- Clone this PR.
- Reinstall Prestashop or execute this script to alter the product table.
```sql
ALTER TABLE `ps_product` CHANGE `redirect_type` `redirect_type` ENUM('','404','410','301-product','302-product','301-category','302-category','200-displayed','404-displayed','410-displayed','default') CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci NOT NULL DEFAULT 'default';
ALTER TABLE `ps_product_shop` CHANGE `redirect_type` `redirect_type` ENUM('','404','410','301-product','302-product','301-category','302-category','200-displayed','404-displayed','410-displayed','default') CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci NOT NULL DEFAULT 'default';
```
- Choose a product, edit it and set product redirect type to `Default`. Visit this product in FO. Disable this product in BO. Keep page open.
- Go to `Shop settings > Product settings` and try to play around with default option. See that if you go to FO and hit F5, the product respects it. You can see the response (200/404/410) if you open F12 in Chrome, go to network tab and click the request.
- Then try setting a specific redirect type on the product page, see that it takes priority over the default one set in prestashop settings.
- Then try going to phpadmin and try to change `redirect_type` to `''` in both `ps_product` and `ps_product_shop` for this product. See that it still works fine and prestashop will respect the default value set in configuration.

### CODE REVIEW

File: src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php]:
```suggestion
                    'You can specify the default behavior for all disabled products. The product page can display a "Discontinued" message or an error page. You can also specify which HTTP response is sent to users. This can be set specifically for each product.',
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
            $this->trans('Default = As specified in the Prestashop settings.', 'Admin.Catalog.Help'),
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
                    $this->trans('No redirection (200), display product', 'Admin.Catalog.Feature') => RedirectType::TYPE_SUCCESS_DISPLAYED,
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
                    $this->trans('No redirection (404), display product', 'Admin.Catalog.Feature') => RedirectType::TYPE_NOT_FOUND_DISPLAYED,
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
                    $this->trans('No redirection (410), display product', 'Admin.Catalog.Feature') => RedirectType::TYPE_GONE_DISPLAYED,
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
                    $this->trans('No redirection (404), display error page', 'Admin.Catalog.Feature') => RedirectType::TYPE_NOT_FOUND,
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
                    $this->trans('No redirection (410), display error page', 'Admin.Catalog.Feature') => RedirectType::TYPE_GONE,
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
            $this->trans('No redirection (200), display product = Do not redirect anywhere, display product as discontinued and return normal 200 response.', 'Admin.Catalog.Help'),
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
            $this->trans('No redirection (404), display product = Do not redirect anywhere, display product as discontinued and return 404 "Not Found" response.', 'Admin.Catalog.Help'),
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
            $this->trans('No redirection (410), display product = Do not redirect anywhere, display product as discontinued and return 410 "Gone" response.', 'Admin.Catalog.Help'),
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
            $this->trans('No redirection (404), display error page = Do not redirect anywhere and display a 404 "Not Found" page.', 'Admin.Catalog.Help'),
```

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
            $this->trans('No redirection (410), display error page = Do not redirect anywhere and display a 410 "Gone" page.', 'Admin.Catalog.Help'),
```

**Review by l-delin — COMMENTED** [review]
edited the wording (consistency) :)

**Comment by Hlavtox** [comment]
@l-delin All solved ;-)

**Comment by Hlavtox** [comment]
@PrestaShop/prestashop-maintainers Big boy ready for review

@jf-viguier Could you also have a look bro? :-)

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (zuk3975) [inline src/Adapter/Presenter/Product/ProductLazyArray.php]: the code doesn't seem to have any relation with the comment?
"discontinued and displayed" is mentioned, but the code only checks if product is active :thinking:

Comment (Hlavtox) [reply src/Adapter/Presenter/Product/ProductLazyArray.php]: What I meant was - if the product is inactive and you chose the redirect option "keep displayed"

File: install-dev/data/db_structure.sql
Comment (zuk3975) [inline install-dev/data/db_structure.sql:1701]: where do we usually put the new configuration values? I think the "PS_PRODUCT_REDIRECTION_DEFAULT" configuration should be saved somewhere with the same value that was default before (404 i think?)

File: src/Core/Domain/Product/ValueObject/RedirectType.php
Comment (zuk3975) [inline src/Core/Domain/Product/ValueObject/RedirectType.php:46]: Do we really need this type?
It sounds strange to return 200, but discontinued. The constant name also sounds strange for same reason :thinking:

File: install-dev/data/db_structure.sql
Comment (Hlavtox) [reply install-dev/data/db_structure.sql:1701]: I don't understand the question. 😄

This is configuration per single product. If product is using `default` as `redirect_type`, prestashop will check `PS_PRODUCT_REDIRECTION_DEFAULT` stored in `ps_configuration`.

File: src/Core/Domain/Product/ValueObject/RedirectType.php
Comment (Hlavtox) [reply src/Core/Domain/Product/ValueObject/RedirectType.php:46]: I have checked how the most successful ecommerce websites do this and many of them are returning normal 200 for discontinued products. 

I think some merchants want to preserve the positions in search engines, even after discontinuing the products. For example if you manufacture something and you still want people to easily find manuals and information about old products. If we offered 404 and 410 only, google would remove it from the results. 

Or, sometimes it can be still better for people to realise that product is discontinued on YOUR PAGE specifically, and choose something current from YOUR offers. If they realise it on your competition's page, you lost a customer. 👍

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (zuk3975) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]: I think this one needs wording modified?
Get rid of "=" sign. And the "Prestashop settings" sounds too abstract, should mention the actual related settings page like "ShopParameters > Product Settings" (or wherever the default is set) or so... ping @Julievrz

Comment (Hlavtox) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]: Well, the = sign was in the original help texts, I followed the example - `Permanent redirection (301) = Permanently display another product or category instead.`

What about `Default = As specified in Shop Parameters > Product settings.`

File: install-dev/data/db_structure.sql
Comment (zuk3975) [inline install-dev/data/db_structure.sql:1699]: I also wonder maybe instead of introducing new enum "default" we could use the default setting when this value is empty string? What happens now if there is empty string vs the "default" ?

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (zuk3975) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]: sounds good to me :man_shrugging: Lets wait for wording experts on this one :v:

File: install-dev/data/db_structure.sql
Comment (zuk3975) [reply install-dev/data/db_structure.sql:1701]: Sorry for bad formulation of the question. I remembered where the default configs are installed.
So,
I think you should add the new `PS_PRODUCT_REDIRECTION_DEFAULT` into
`install-dev/data/xml/configuration.xml`
so that the new config value exists after installation by default.

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (zuk3975) [reply src/Adapter/Presenter/Product/ProductLazyArray.php]: yeah but by looking at the code here is nothing related to the redirection, it only disables the button if product is inactive, so I think the comment should be removed or changed to smth like `if the product is disabled, we disable the addToCartButton too` (but I think its self explanatory, so you can just get rid of the comment)

File: install-dev/data/db_structure.sql
Comment (Hlavtox) [reply install-dev/data/db_structure.sql:1701]: Aaaah, now I understand. :-) Forgot it, will do!

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (Hlavtox) [reply src/Adapter/Presenter/Product/ProductLazyArray.php]: OK! Good idea. :-)

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (Hlavtox) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]: @l-delin @Julievrz We need you! 😄

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (zuk3975) [inline src/Adapter/Presenter/Product/ProductLazyArray.php]: just a minor suggestion (only my opinion :man_shrugging: ). I prefer using same keywords in code (I think everyone in prestashop are familiar with the keywords "disabled|enabled or "active|inactive", but not yet so familiar with the new keyword you just introduced "discontinued". Furthermore, I think product can be inactive for multiple reasons, not only discontinued, lets say temporarily disabled or so...). So the following should be more appropriate and easier to understand to anyone.

```suggestion
            // If the product is disabled, but still displayed, we display a proper message
```

File: install-dev/data/db_structure.sql
Comment (Hlavtox) [reply install-dev/data/db_structure.sql:1699]: The current behavior (that I retained kept), is that if redirect_type ends up empty/null/false, it will use classic 404 page+response.

Comment (Hlavtox) [reply install-dev/data/db_structure.sql:1699]: Also, just realised there is '' in shop table, should I also remove it like you did here? :-) https://github.com/PrestaShop/PrestaShop/pull/22378

I will make the default value 'default' in both cases.

Comment (zuk3975) [reply install-dev/data/db_structure.sql:1699]: in that case we wouldn't need to add additional "default" value in the RedirectType value object, but when loading the product we would have to check if the type is not set (if its the "empty string" one), then load the appropriate RedirectType based on configuration value.
That would also allow you to avoid chaning the default "404" redirect type in behats by instead setting up the default common configuration value `PS_PRODUCT_REDIRECTION_DEFAULT` to a 404.

though ofc I haven't considered this thoroughly, as I didin't dive too deep into those scenarios, so its just a suggestion to think about (maybe its a bad idea, but maybe good? :smile: idk :man_shrugging:)

Comment (zuk3975) [reply install-dev/data/db_structure.sql:1699]: `Also, just realised there is '' in shop table, should I also remove it like you did here?`
Instead of removing it you can leave it empty and get rid of the "default" one. This way just avoiding the big BC break. (therefore the empty string would just act as a 'default' just ofc it wouldn't be so verbal anymore) :man_shrugging:

File: src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php
Comment (l-delin) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]:
```suggestion
            $this->trans('Default behavior is specified in the Shop Parameters > Product settings page.', 'Admin.Catalog.Help'),
```

Comment (l-delin) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/SEO/RedirectOptionType.php]: Hello, we reviewed the wording 🔧👏 @Hlavtox

File: js/admin.js
Comment (zuk3975) [inline js/admin.js:544]: I see these used in couple place already, can't we use some constants or smth, to avoid repeating and more important - avoid having magic strings|numbers?
Though, Im not the expert of js, especially in this old code part, maybe @NeOMakinG can elaborate?

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (zuk3975) [inline src/Adapter/Presenter/Product/ProductLazyArray.php:392]: Also just a thought (not required), but shouldn't it be better to put these values into some constants (at least here in the ProductLazyArray if they are not used anywhere else for now)? Its not recommended having plain strings hanging.

**Review by zuk3975 — DISMISSED** [review]
ok for me, except some suggestions to add constants

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (kpodemski) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:392]: src/Core/Domain/Product/SeoAvailability? just like ProductSettings

Comment (zuk3975) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:392]: hmm that wouldn't fit into a ValueObject definition, because its not something we will initialize, it will only hold the constants. (ValueObjects should be initializable and validate itself  - so a good example of value object would be if we need something like `new SeoAvailability($value)`. But in this case we won't ever initialize the class. I don't have any other suggestions where to put it though... Maybe @jolelievre have any thoughts?

Comment (Hlavtox) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:392]: Guys is this really needed? It's one place where this is used. Do we need to test that if I create new SeoAvailability it will return the string I hardcoded there? ... come on

Comment (kpodemski) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:392]: @Hlavtox it's easier to maintain this way, if they ever change Available to "SomethingElse" you change the value in the constant and the rest of the system works as before

@zuk3975 I edited my message right after I wrote about VO ;)

Comment (zuk3975) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:392]: @kpodemski Yeah probably sounds good. Though it would need comments above the class about what it represents (and a private constructor, so it cannot be initialized), so we don't confuse it with a value object.

Comment (Hlavtox) [reply src/Adapter/Presenter/Product/ProductLazyArray.php:392]: I have other PR that modifies this, so let's do this after it's merged pls ;-) https://github.com/PrestaShop/PrestaShop/pull/29570

**Review by kpodemski — DISMISSED** [review]

**Comment by kpodemski** [comment]
Well done, @Hlavtox 

Let's try to merge it first, it can be improved a bit later. I've added "Key feature" label so we can remember to highlight this change during the Public Demo 😎

File: admin-dev/themes/default/js/bundle/product/form.js
Comment (NeOMakinG) [inline admin-dev/themes/default/js/bundle/product/form.js:2222]: Nitpick but you could make it a bit less verbose and functional by doing

```
const redirectTypeValues = [
  '410',
  '404',
  'default',
  '200-displayed',
  '404-displayed',
  '410-displayed',
]

if (redirectTypeValues.includes(redirectTypeValue) {
```

Comment (Hlavtox) [reply admin-dev/themes/default/js/bundle/product/form.js:2222]: @NeOMakinG It's already in QA dude I don't want to ping everyone again. 😄

**Comment by hibatallahAouadni** [comment]
@Hlavtox could you please resolve the conflicts :pray:

**Comment by Hlavtox** [comment]
@hibatallahAouadni rebased

**Review by zuk3975 — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Review by hibatallahAouadni — COMMENTED** [review]
Hello @Hlavtox 

With you PR, everything works perfectly except the redirection 302 :disappointed: 
When I choose the redirection `Temporary redirection (302)` it will be displayed in the console of FO as `Permanent redirection (301)`, see the attached screen record below:

[pr_29720.webm](https://user-images.githubusercontent.com/23308427/203087149-184c83e2-e5e0-4fee-9517-84d8397d0ef2.webm)

See in develop, it works:
image
Waiting for your feedback.

Thanks!

**Review by marsaldev — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]
Thanks @Hlavtox

**Comment by zuk3975** [comment]
sorry, what is the state of PR?
(QA waited for author feedback, but label was taken off without any mention if the question been resolved or not)

**Comment by Hlavtox** [comment]
@zuk3975 

@hibatallahAouadni secretly made a QA ✅, but can't put a comment here because the PR is not reapproved after rebase. 😄

And that previous comment is just a browser cache issue that was resolved, but also not commented or comment deleted.

**Comment by hibatallahAouadni** [comment]
Hello @zuk3975 

As mentioned by @Hlavtox my previous comment was due to cache issue.
Sorry 😞

**Review by hibatallahAouadni — APPROVED** [review]
Hello @Hlavtox 

LGTM, QA ✅ 
It works perfectly with each redirect status 👏 

Thanks!

File: classes/Product.php
Comment (FabienPapet) [inline classes/Product.php:215]: Looks like a huge BC break here :/

Comment (zuk3975) [reply classes/Product.php:215]: not as huge as it looks (or not a BC at all?). because the TYPE_DEFAULT will contain the same value as it was under "TYPE_NOT_FOUND" by default in configuration as you can see [here](https://github.com/PrestaShop/PrestaShop/pull/29720/files#diff-d9529f3253ec6cb566b81d6c6a5943efab9fbdd1db543b99ca4cdbbc9995dfdfR895-R896).

Comment (Hlavtox) [reply classes/Product.php:215]: @FabienPapet What do you have in mind? This won't break anything. No value was deleted from RedirectType class. Also, whatever value there willl be in database, it will fallback both in frontoffice and backoffice.

Comment (Hlavtox) [reply classes/Product.php:215]: Discussed with Jo and Krystain, no BC break ;-)

**Comment by zuk3975** [comment]
It takes ages for build, something is wrong. I close and reopen PR to restart the build

File: classes/Product.php
Comment (jolelievre) [reply classes/Product.php:215]: Yep, not a BC it changes the initial behaviour indeed but it states that it should use the global default configuration, and the global default value is the same 404 initially. You can change it afterward but initially, nothing changes.

**Pull request merged**
